File: sharedse/src/main/java/com/ustadmobile/port/sharedse/networkmanager/DownloadJobPreparer.java
package com.ustadmobile.port.sharedse.networkmanager;

import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.db.dao.DownloadJobItemDao;
import com.ustadmobile.core.impl.UMLog;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;
import com.ustadmobile.core.networkmanager.DownloadJobItemManager;
import com.ustadmobile.lib.db.entities.Container;
import com.ustadmobile.lib.db.entities.DownloadJobItem;
import com.ustadmobile.lib.db.entities.DownloadJobItemParentChildJoin;
import com.ustadmobile.lib.util.UMUtil;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

/**
 * This runnable sets up a download job so it's ready to run. It starts from a root content entry uid,
 * and then adds all
 */
public class DownloadJobPreparer implements Runnable {

    private DownloadJobItemManager jobItemManager;

    private UmAppDatabase appDatabase;

    private UmAppDatabase appDatabaseRepo;

    public DownloadJobPreparer(DownloadJobItemManager jobItemManager, UmAppDatabase appDatabase,
                               UmAppDatabase appDatabaseRepo) {
        this.jobItemManager = jobItemManager;
        this.appDatabase = appDatabase;
        this.appDatabaseRepo = appDatabaseRepo;
    }

    @Override
    public void run() {
        long startTime = System.currentTimeMillis();
        int downloadJobUid = jobItemManager.getDownloadJobUid();
        long contentEntryUid = jobItemManager.getRootContentEntryUid();

        UstadMobileSystemImpl.l(UMLog.DEBUG, 420, "DownloadJobPreparer: start " +
                "entry uid = " + contentEntryUid + " download job uid = " + downloadJobUid);

        int numItemsCreated = 0;

        DownloadJobItemDao jobItemDao = appDatabase.getDownloadJobItemDao();
        List<DownloadJobItemDao.DownloadJobItemToBeCreated2> childItemsToCreate;
        Container rootEntryContainer = appDatabaseRepo.getContainerDao()
                .getMostRecentContainerForContentEntry(contentEntryUid);
        DownloadJobItem rootDownlaodJobItem = new DownloadJobItem(
                jobItemManager.getDownloadJobUid(), contentEntryUid,
                rootEntryContainer != null ? rootEntryContainer.getContainerUid() : 0,
                rootEntryContainer != null ? rootEntryContainer.getFileSize() : 0);
        jobItemManager.insertDownloadJobItemsSync(Collections.singletonList(rootDownlaodJobItem));

        Map<Long, Long> contentEntryUidToDjiUidMap = new HashMap<>();
        List<Long> parentUids = new ArrayList<>();
        parentUids.add(contentEntryUid);
        contentEntryUidToDjiUidMap.put(contentEntryUid, rootDownlaodJobItem.getDjiUid());

        HashSet<Long> createdJoinCepjUids = new HashSet<>();
        do {
            childItemsToCreate = jobItemDao.findByParentContentEntryUuids(parentUids);
            UstadMobileSystemImpl.l(UMLog.DEBUG, 420, "DownloadJobPreparer: found " +
                    childItemsToCreate.size() + " child items on from parents " +
                    UMUtil.debugPrintList(parentUids));

            parentUids.clear();

            for(DownloadJobItemDao.DownloadJobItemToBeCreated2 child : childItemsToCreate){
                if(!contentEntryUidToDjiUidMap.containsKey(child.getContentEntryUid())) {
                    DownloadJobItem newItem = new DownloadJobItem(downloadJobUid,
                            child.getContentEntryUid(), child.getContainerUid(), child.getFileSize());
                    jobItemManager.insertDownloadJobItemsSync(Arrays.asList(newItem));
                    numItemsCreated++;

                    contentEntryUidToDjiUidMap.put(child.getContentEntryUid(),
                            newItem.getDjiUid());

                    if(newItem.getDjiContainerUid() == 0) //this item is a branch, not a leaf if containeruid = 0
                        parentUids.add(child.getContentEntryUid());

                }

                if(!createdJoinCepjUids.contains(child.getCepcjUid())) {
                    jobItemManager.insertParentChildJoins(Arrays.asList(new DownloadJobItemParentChildJoin(
                            contentEntryUidToDjiUidMap.get(child.getParentEntryUid()),
                            contentEntryUidToDjiUidMap.get(child.getContentEntryUid()),
                            child.getCepcjUid())
                    ), null);
                    createdJoinCepjUids.add(child.getCepcjUid());
                }
            }

        }while(!parentUids.isEmpty());
        UstadMobileSystemImpl.l(UMLog.VERBOSE, 420, "Created " + numItemsCreated +
                " items. Time to prepare download job: " +
                (System.currentTimeMillis() - startTime) + "ms");
        CountDownLatch latch = new CountDownLatch(1);
        jobItemManager.commit((aVoid) -> latch.countDown());
        try { latch.await(5, TimeUnit.SECONDS); }
        catch(InterruptedException e) {}
    }
}

File: sharedse/src/commonJvmMain/kotlin/com/ustadmobile/port/sharedse/networkmanager/DownloadJobPreparer.kt
package com.ustadmobile.port.sharedse.networkmanager

import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.db.dao.DownloadJobItemDao
import com.ustadmobile.core.impl.UMLog
import com.ustadmobile.core.impl.UmResultCallback
import com.ustadmobile.lib.db.entities.ContentEntryStatus
import com.ustadmobile.lib.db.entities.DownloadJobItem
import com.ustadmobile.lib.db.entities.DownloadJobItemParentChildJoin
import com.ustadmobile.lib.util.UMUtil
import com.ustadmobile.sharedse.network.DownloadJobItemManager
import java.util.*
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * This runnable sets up a download job so it's ready to run. It starts from a root content entry uid,
 * and then adds all
 */
class DownloadJobPreparer(private val jobItemManager: DownloadJobItemManager, private val appDatabase: UmAppDatabase,
                          private val appDatabaseRepo: UmAppDatabase) : Runnable {

    override fun run() {
        val startTime = System.currentTimeMillis()
        val downloadJobUid = jobItemManager.downloadJobUid
        val contentEntryUid = jobItemManager.rootContentEntryUid

        UMLog.l(UMLog.DEBUG, 420, "DownloadJobPreparer: start " +
                "entry uid = " + contentEntryUid + " download job uid = " + downloadJobUid)

        var numItemsCreated = 0

        val jobItemDao = appDatabase.downloadJobItemDao
        var childItemsToCreate: List<DownloadJobItemDao.DownloadJobItemToBeCreated2>
        val rootEntryContainer = appDatabaseRepo.containerDao
                .getMostRecentContainerForContentEntry(contentEntryUid)
        val rootDownlaodJobItem = DownloadJobItem(
                jobItemManager.downloadJobUid, contentEntryUid,
                rootEntryContainer?.containerUid ?: 0,
                rootEntryContainer?.fileSize ?: 0)
        TODO("Make this use coroutine suspended function")
        //jobItemManager.insertDownloadJobItemsSync(listOf(rootDownlaodJobItem))

        val contentEntryUidToDjiUidMap = HashMap<Long, Int>()
        val parentUids = ArrayList<Long>()
        parentUids.add(contentEntryUid)
        contentEntryUidToDjiUidMap[contentEntryUid] = rootDownlaodJobItem.djiUid

        val createdJoinCepjUids = HashSet<Long>()

        appDatabase.contentEntryStatusDao.insertOrAbort(listOf(ContentEntryStatus(contentEntryUid,
                rootEntryContainer == null,
                rootEntryContainer?.fileSize ?: 0)))

        val statusList = LinkedList<ContentEntryStatus>()
        do {
            statusList.clear()
            childItemsToCreate = jobItemDao.findByParentContentEntryUuids(parentUids)
            UMLog.l(UMLog.DEBUG, 420, "DownloadJobPreparer: found " +
                    childItemsToCreate.size + " child items on from parents " +
                    UMUtil.debugPrintList(parentUids))

            parentUids.clear()

            for (child in childItemsToCreate) {
                if (!contentEntryUidToDjiUidMap.containsKey(child.contentEntryUid)) {
                    val newItem = DownloadJobItem(downloadJobUid,
                            child.contentEntryUid, child.containerUid, child.fileSize)
                    TODO("Make this use suspend function")
                    //jobItemManager.insertDownloadJobItemsSync(listOf(newItem))
                    numItemsCreated++
                    statusList.add(ContentEntryStatus(child.contentEntryUid,
                            child.fileSize > 0, child.fileSize))

                    contentEntryUidToDjiUidMap[child.contentEntryUid] = newItem.djiUid

                    if (newItem.djiContainerUid == 0L)
                    //this item is a branch, not a leaf if containeruid = 0
                        parentUids.add(child.contentEntryUid)

                    statusList.add(ContentEntryStatus(child.contentEntryUid,
                            child.fileSize > 0, child.fileSize))

                }

                if (!createdJoinCepjUids.contains(child.cepcjUid)) {
                    TODO("Make this use suspend function")
//                    jobItemManager.insertParentChildJoins(listOf(DownloadJobItemParentChildJoin(
//                            contentEntryUidToDjiUidMap[child.parentEntryUid]!!,
//                            contentEntryUidToDjiUidMap[child.contentEntryUid]!!,
//                            child.cepcjUid)), null)
                    createdJoinCepjUids.add(child.cepcjUid)
                }
            }

            appDatabase.contentEntryStatusDao.insertOrAbort(statusList)
        } while (!parentUids.isEmpty())
        UMLog.l(UMLog.VERBOSE, 420, "Created " + numItemsCreated +
                " items. Time to prepare download job: " +
                (System.currentTimeMillis() - startTime) + "ms")
        val latch = CountDownLatch(1)
        TODO("Fix this to use coroutine suspend function")
//        jobItemManager.commit(object: UmResultCallback<Void>{
//            override fun onDone(result: Void?) {
//                latch.countDown()
//            }
//
//        })
        try {
            latch.await(5, TimeUnit.SECONDS)
        } catch (e: InterruptedException) {
        }

    }
}
